### PR TITLE
Module 10

### PR DESCRIPTION
**Pull request by JoeBawn**
Please make sure you are raising this PR against your own repository and not the original. If it says https://github.com/CorndelWithSoftwire in the search bar right now, you are in the wrong place!

Please also don't forget to submit the PR's url to Aptem afterwards (by clicking on the corresponding exercise component, pasting the url in the textbox that appears and then pressing the finish button).

### CODE REVIEW

File: todo_app/app.py
Comment (JackMead) [inline todo_app/app.py:52]: Currently they is checking _the string_ "current_user" so it always comes up false!
```suggestion
        if hasattr(current_user, 'role'):
```

File: todo_app/app.py
Comment (JackMead) [inline todo_app/app.py:57]: Is defaulting to the write-template sensible behaviour? If a new role is introduced and someone forgets to update this code, then we're going to risk exposing functionality that they shouldn't have

File: todo_app/app.py
Comment (JackMead) [inline todo_app/app.py:60]: I suspect this is meant to deal with the case for testing? If so I'd be inclined to program this a bit more defensively to make that obvious and rather than just letting it fall back to this if the user object has its role removed; e.g. checking if the LOGIN_DISABLED config is set and then returning the write template, otherwise checking the role?

File: todo_app/app.py
Comment (JackMead) [inline todo_app/app.py:29]: This will work here but would be better as an environment variable, otherwise:
* If we tried to scale the server by running multiple instances, they'll all have different secrets and users will lose their session if they end up talking to different servers
* Every time the server is restarted all sessions will be invalidated

File: todo_app/app.py
Comment (JackMead) [inline todo_app/app.py:64]: We should also add some sort of role check within each of these functions (which we could do with a custom decorator or by checking directly inside each block). As they currently stand, it's very easy for readers to nefariously get write access:
* If they're accidentally served the wrong template then there's no protection
* If they know/guess the appropriate endpoint then they can bypass needing the right template
We definitely need to have role checks implemented server side for appropriate functions

File: todo_app/templates/index_read.html
Comment (JackMead) [inline todo_app/templates/index_read.html:1]: Nothing wrong with this approach, but it looks like a fair amount of duplication - might be worth considering whether it would be easier (even if not worth switching) to pass the role through as part of the view model, and conditionally render certain sections

**Comment by JoeBawn** [comment]
Hi Jack, 

I have gone through and made the majority of the changes you recommended. Can you approve when you are happy then I will merge. 

Cheers
Joe

File: todo_app/app.py
Comment (JackMead) [inline todo_app/app.py:134]: I missed this last time as the authorisation wasn't hiding anything, but I don't think your app ever treats anyone as a user - the "id" field passed through here (`github_user`) is actually a large json object with lots of fields, so the below never passes
```python
if self.id == 'JoeBawn':
            return 'writer'
```

Instead, we want to just read a specific field out of this response;
```suggestion
        login_user(User(github_user['login']))
```

File: todo_app/app.py
Comment (JackMead) [inline todo_app/app.py:30]: Nice idea, but you always set the value of LOGIN_DISABLED here, even if you're sometimes setting it to false, but your code expects the value to be empty if it's not true! As a read only user I still saw the write page, and was able to submit things

Instead you could either:
* Adjust the checks to be more like: `if not app.config['LOGIN_DISABLED']:`
* Or only set the app config if it's true:
```suggestion
    if os.environ.get('LOGIN_DISABLED', 'False').lower() in ['true', '1']:
        app.config['LOGIN_DISABLED'] = True
```

**Review by JackMead — APPROVED** [review]
Nice work and thanks for dealing with those - I've left a couple more comments below because I can see the structure is mostly in place but you've still got a couple of bugs such that read-only users can still access your endpoints & no user is given write access! It's only small fixes so I'm happy to approve this but do make those changes and test they're working before you merge. If it's not clear what's wrong or what the solutions are then do reach out 🙂

**Comment by JoeBawn** [comment]
Thanks Jack, I have made the suggested changes and merged.

**Pull request merged**
